splice_si_api/src/main/java/com/splicemachine/si/impl/server/SITransactor.java

```diff
@@ -298,33 +300,64 @@ private void releaseLocksForKvBatch(Pair<KVPair, Lock>[] locks){
                  */
                 //todo -sf remove the Row key copy here
                 checkedConflicts = true;
-                possibleConflicts=bloomInMemoryCheck==null||bloomInMemoryCheck.get(i)?table.getLatest(kvPair.getRowKey(),possibleConflicts):null;
-                if(possibleConflicts!=null && !possibleConflicts.isEmpty()){
-                    //we need to check for write conflicts
-                    try {
-                        conflictResults = ensureNoWriteConflict(transaction, writeType, possibleConflicts, rollForwardQueue, supplier);
-                    } catch (IOException ioe) {
-                        if (ioe instanceof WriteConflict) {
-                            finalStatus[i] = operationStatusLib.failure(ioe);
-                            continue;
-                        } else throw ioe;
+                boolean shouldContinue = true;
+                while(shouldContinue) {
+                    possibleConflicts = bloomInMemoryCheck == null ||
+                            bloomInMemoryCheck.get(i) ? table.getLatest(kvPair.getRowKey(), possibleConflicts) : null;
+                    if (possibleConflicts != null && !possibleConflicts.isEmpty()) {
+                        if (LOG.isDebugEnabled()) {
+                            for (DataCell dataCell : possibleConflicts) {
+                                CellType type = dataCell.dataType();
+                                byte[] k = dataCell.key();
+                                byte[] v = dataCell.value();
+                                SpliceLogUtils.debug(LOG, "Possible conflicts: %s, k%s, v=%s",
+                                        type.toString(), Bytes.toStringBinary(k), Bytes.toStringBinary(v));
+                            }
+                        }
+                        //we need to check for write conflicts
+                        try {
+                            conflictResults = ensureNoWriteConflict(transaction, writeType, possibleConflicts, rollForwardQueue, supplier);
+                        } catch (IOException ioe) {
+                            if (ioe instanceof WriteConflict) {
+                                finalStatus[i] = operationStatusLib.failure(ioe);
+                                shouldFail = true;
+                                break;
+                            } else throw ioe;
+                        }
+
+                        if (constraintChecker != null) {
+                            List<DataCell> visibleColumns = filterCells(possibleConflicts, constraintStateFilter);
+                            if (visibleColumns.size() == 0 && !conflictResults.hasAdditiveConflicts()) {
```

Comment:
Given that we already filtered out the Rolledback txns I think we don't need to filter them out again. If some active txn just became rolledback, so be it, raising a WW conflict (or PK violation) is fine.

Reply:
That will report some false positives

Reply:
I don't understand this, if we've already filtered out rolledback rows we don't need to check again. Do you have an example for those false positives?

Reply:
It could happen that the latest row returned was rolled back after it was returned. Either we report a WW conflict or get next row that is visible to check WW conflict.